icicle_v3/backend/cpu/include/cpu_ntt.h

```diff
@@ -0,0 +1,331 @@
+    }
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename S>
+  eIcicleError CpuNttDomain<S>::cpu_ntt_release_domain(const Device& device)
+  {
+    // No need to explicitly free memory since unique_ptr will handle it automatically
+    s_ntt_domain.twiddles.reset(); // Set twiddles to nullptr
+    s_ntt_domain.max_size = 0;
+    s_ntt_domain.max_log_size = 0;
+    return eIcicleError::SUCCESS;
+  }
+
+  int bit_reverse(int n, int logn)
+  {
+    int rev = 0;
+    for (int j = 0; j < logn; ++j) {
+      if (n & (1 << j)) { rev |= 1 << (logn - 1 - j); }
+    }
+    return rev;
+  }
+
+  template <typename E = scalar_t>
+  eIcicleError reorder_by_bit_reverse(int logn, E* output, int batch_size)
+  {
+    int size = 1 << logn;
+    for (int batch = 0; batch < batch_size; ++batch) {
+      E* current_output = output + batch * size;
+      int rev;
+      for (int i = 0; i < size; ++i) {
+        rev = bit_reverse(i, logn);
+        if (i < rev) { std::swap(current_output[i], current_output[rev]); }
+      }
+    }
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename S = scalar_t, typename E = scalar_t>
+  void dit_ntt(E* elements, int size, int batch_size, const S* twiddles, NTTDir dir, int domain_max_size)
+  {
+    for (int batch = 0; batch < batch_size; ++batch) {
+      E* current_elements = elements + batch * size;
+      for (int len = 2; len <= size; len <<= 1) {
+        int half_len = len / 2;
+        int step = (size / len) * (domain_max_size / size);
+        for (int i = 0; i < size; i += len) {
+          for (int j = 0; j < half_len; ++j) {
+            int tw_idx = (dir == NTTDir::kForward) ? j * step : domain_max_size - j * step;
+            E u = current_elements[i + j];
+            E v = current_elements[i + j + half_len] * twiddles[tw_idx];
+            current_elements[i + j] = u + v;
+            current_elements[i + j + half_len] = u - v;
+          }
+        }
+      }
+    }
+  }
+
+  template <typename S = scalar_t, typename E = scalar_t>
+  void dif_ntt(E* elements, int size, int batch_size, const S* twiddles, NTTDir dir, int domain_max_size)
+  {
+    for (int batch = 0; batch < batch_size; ++batch) {
+      E* current_elements = elements + batch * size;
+      for (int len = size; len >= 2; len >>= 1) {
+        int half_len = len / 2;
+        int step = (size / len) * (domain_max_size / size);
+        for (int i = 0; i < size; i += len) {
+          for (int j = 0; j < half_len; ++j) {
+            int tw_idx = (dir == NTTDir::kForward) ? j * step : domain_max_size - j * step;
+            E u = current_elements[i + j];
+            E v = current_elements[i + j + half_len];
+            current_elements[i + j] = u + v;
+            current_elements[i + j + half_len] = (u - v) * twiddles[tw_idx];
+          }
+        }
+      }
+    }
+  }
+
+  template <typename E = scalar_t>
+  void transpose(const E* input, E* output, int rows, int cols)
```

Comment:
note that there is a transpose api in vec_ops.h you can use if you need.
Also it's true for bit-reverse BTW.
That way you don't need to implement it again.